### ISSUE TITLE
beta sometimes returns NaN

### ISSUE THREAD

**Issue by longouyang** [probmods/webppl#269]
( I'm on the dev branch -- 55e342a0393421ce1a87eda36211b8c7d8b057be -- although this bug is present on the current webppl.org version as well; I suspect that it's been with us since the very beginning )

Here's a fairly simple model:
```js
repeat(10, function() { return beta(0.001,0.001) });
```

With some troubling output:
```js
[ 1,
  1,
  9.235726936878123e-267,
  NaN,
  0,
  0,
  1,
  NaN,
  2.26873135676291e-25,
  NaN ]
```

For context, the beta sampler is:
```js
function betaSample(params) {
  var a = params[0];
  var b = params[1];
  var x = gammaSample([a, 1]);
  return x / (x + gammaSample([b, 1]));
}
```

The cause of the NaN error is that `x` and `x + gammaSample([b,1])` can both be 0 (0/0 gives NaN in Javascript). A simple patch would be:

```js
function betaSample(params) {
  var a = params[0];
  var b = params[1];
  var x = 0, y = 0;
  // if we hit x == 0 and y == 0, resample
  while (x == 0 && x + y == 0) {
    x = gammaSample([a, 1]);
    y = gammaSample([b, 1]);
  }
  return x / (x + y);
}
```

(in my informal testing, you can't just treat 0/0 as 1; it doesn't sample things in the right probabilities).
I'm hesitant to apply this fix because the better solution is to prevent `gammaSample` from returning 0, which I will soon work on.

More generally, I'm kind of alarmed at the lack of documentation in erp.js (e.g., which papers / websites do our implementations come from?) as well as the paucity of statistical sanity checks in the unit tests.

**Comment by null-a** [comment]
> the better solution is to prevent gammaSample from returning 0, which I will soon work on.

Great. You've probably seen it, but for reference that is issue #132.

**Comment by stuhlmueller** [comment]
A similar issue sometimes afflicts dirichlet:

    dirichlet([.0001, .0001])
    # [null,null]

This came up in [this mailing list discussion](https://groups.google.com/forum/#!topic/webppl-dev/jTi4INaFDfQ) where MH presented a model that has essentially this structure:

    MH(function(){
      var alpha = dirichlet([.1,.1])
      return dirichlet(alpha)
    }, 10)

**Comment by longouyang** [comment]
Yes, although simply fixing the bug in gamma does not completely solve the dirichlet bug because we can still underflow when we normalize the sampled gamma values:

```js
  for (var i = 0; i < alpha.length; i++) {
    t = gammaSample([alpha[i], 1]); // can underflow
    theta[i] = t;
    ssum = ssum + t;
  }
  for (var j = 0; j < theta.length; j++) {
    theta[j] /= ssum;               // can underflow
  }
```

I have proposed a [trick for mitigating underflow in issue 132](https://github.com/probmods/webppl/issues/132#issuecomment-156790566) -- can you provide feedback?

**Comment by null-a** [comment]
For reference, here's a previous [discussion](https://github.com/probmods/webppl/pull/277#issuecomment-165332720) about fixing the beta/dirichlet samplers using the [`expGamma`](https://github.com/probmods/webppl/blob/ef951a0f434116ef020b283296ed7ade7ad3b3f1/src/erp.ad.js#L330) sampler.

**Comment by longouyang** [comment]
Leaving a paper trail: this got sort of addressed in #514 but that correction leads to incorrect kurtosis for `beta(0.001, 0.001)` (discovered while writing tests for beta sampler) -- #536 fixes this by switching to `expGammaSample` for the Beta sampler.
